### PR TITLE
8283042: Polish java.lang.foreign javadoc

### PR DESCRIPTION
**Pull request by mcimadamore**
This patch addresses the javadoc inconsistencies mentioned in JDK-8283042.

Here's a summary of the changes:

* treatment for `nulls` has been moved to toplevel javadoc
* third person has been chosen over infinitive form in all method javadoc (this seems to be the most common form in JDK code, after some grepping)
* `Creates` has been used in all places where a "new" object is returned;
* `Returns` has been used in all places where a view of an object is returned
* dropped `new` in all places where it could be over-constraining or misleading; I've also dropped `new` immediately after a `Creates`
* Replaces most instances of `Obtains` with either `Creates` or `Returns` (which seems way more common, after some grepping)
* Made javadoc for *wither* methods in `MemoryLayout` and `FunctionDescriptor` more uniform
* Fix issue where layout subclasses did not inherit `MemoryLayout` javadoc for Object methods (due to javadoc issue)
* Make sure that first sentence of all method javadoc is short and descriptive, and looks good in the class summary javadoc
* Slightly tweaked the first sentence of the top-level package javadoc (to make sure it looks good in the java.base summary
* Other minor fixes, mainly aimed to reduce some of the verbosity.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Issue
 * [JDK-8283042](https://bugs.openjdk.java.net/browse/JDK-8283042): Polish java.lang.foreign javadoc

### Reviewers
 * [Julia Boes](https://openjdk.java.net/census#jboes) (@FrauBoes - Committer) ⚠️ Review applies to f31c9483cbcfd985daed0d4dce426435eb0f73d1

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/panama-foreign pull/669/head:pull/669` \
`$ git checkout pull/669`

Update a local copy of the PR: \
`$ git checkout pull/669` \
`$ git pull https://git.openjdk.java.net/panama-foreign pull/669/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 669`

View PR using the GUI difftool: \
`$ git pr show -t 669`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/panama-foreign/pull/669.diff">https://git.openjdk.java.net/panama-foreign/pull/669.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/lang/foreign/AbstractLayout.java [hunk 1]
```diff
@@ -144,25 +144,49 @@ public boolean isPadding() {
...
 
+    /**
+     * {@return the hash code value for this layout}
+     */
     @Override
     public int hashCode() {
         return name.hashCode() << Long.hashCode(alignment);
     }
 
+    /**
```
Comment (mcimadamore) [inline src/java.base/share/classes/java/lang/foreign/AbstractLayout.java:155]: This, and other javadoc for Object methods has to be copied from `MemoryLayout` to here, or the javadoc rendering for the layout subclasses will just display the Object method javadoc.

**Comment by mcimadamore** [comment]
Javadoc link available here:
http://cr.openjdk.java.net/~mcimadamore/panama/foreign_preview_javadoc_tweaks/javadoc/java.base/module-summary.html

File: src/java.base/share/classes/java/lang/foreign/CLinker.java [hunk 2]
```diff
@@ -182,10 +181,10 @@ default MethodHandle downcallHandle(Addressable symbol, FunctionDescriptor funct
     }
 
     /**
-     * Obtains a foreign method handle, with the given type and featuring the given function descriptor, which can be
-     * used to call a target foreign function at the address in a dynamically provided {@link Addressable} instance.
+     * Creates a method handle which can be used to call a target foreign function with given signature.
```
Comment (FrauBoes) [inline src/java.base/share/classes/java/lang/foreign/CLinker.java:184]: same, "with **the** given signature"

File: src/java.base/share/classes/java/lang/foreign/CLinker.java [hunk 3]
```diff
@@ -239,9 +237,8 @@ static MethodType downcallType(FunctionDescriptor functionDescriptor) {
     }
 
     /**
-     * Obtains the method handle {@linkplain MethodType type} associated with an upcall stub with given function descriptor.
+     * {@return the method handle {@linkplain MethodType type} associated with an upcall stub with given function descriptor}
```
Comment (FrauBoes) [inline src/java.base/share/classes/java/lang/foreign/CLinker.java:240]: with **the** given function descriptor

File: src/java.base/share/classes/java/lang/foreign/CLinker.java [hunk 4]
```diff
@@ -228,9 +227,8 @@ default MethodHandle downcallHandle(Addressable symbol, FunctionDescriptor funct
     MemorySegment upcallStub(MethodHandle target, FunctionDescriptor function, MemorySession session);
 
     /**
-     * Obtains the downcall method handle {@linkplain MethodType type} associated with a given function descriptor.
+     * {@return the downcall method handle {@linkplain MethodType type} associated with a given function descriptor}
```
Comment (FrauBoes) [inline src/java.base/share/classes/java/lang/foreign/CLinker.java:230]: with **the** given function descriptor

File: src/java.base/share/classes/java/lang/foreign/CLinker.java [hunk 5]
```diff
@@ -194,15 +193,15 @@ default MethodHandle downcallHandle(Addressable symbol, FunctionDescriptor funct
...
+     * @return a new downcall method handle. The method handle type is <a href="CLinker.html#downcall-method-handles"><em>inferred</em></a>
      * from the provided function descriptor.
-     * @throws IllegalArgumentException if the provided descriptor contains either a sequence or a padding layout.
+     * @throws IllegalArgumentException if the provided function descriptor contains either a sequence or a padding layout.
      */
     MethodHandle downcallHandle(FunctionDescriptor function);
 
     /**
-     * Allocates a native stub which can be passed to other foreign functions (as a function pointer), with given
+     * Creates a native stub which can be passed to other foreign functions as a function pointer, with given
```
Comment (FrauBoes) [inline src/java.base/share/classes/java/lang/foreign/CLinker.java:204]: with **the** given memory session

File: src/java.base/share/classes/java/lang/foreign/MemoryAddress.java [hunk 6]
```diff
@@ -27,6 +27,8 @@
 package java.lang.foreign;
 
 import java.nio.ByteOrder;
+
+import jdk.internal.ValueBased;
```
Comment (FrauBoes) [inline src/java.base/share/classes/java/lang/foreign/MemoryAddress.java:31]: Is this an unused import?

File: src/java.base/share/classes/java/lang/foreign/MemorySegment.java [hunk 7]
```diff
@@ -873,8 +870,8 @@ static MemorySegment ofAddress(MemoryAddress address, long bytesSize, MemorySess
     }
 
     /**
-     * Creates a new native memory segment that models a newly allocated block of off-heap memory with given layout
-     * and memory session. A client is responsible make sure that the memory session associated with the returned segment is closed
+     * Creates a native memory segment with given layout and memory session.
+     * A client is responsible make sure that the memory session associated with the returned segment is closed
```
Comment (FrauBoes) [inline src/java.base/share/classes/java/lang/foreign/MemorySegment.java:874]: is responsible **for** ensuring that

File: src/java.base/share/classes/java/lang/foreign/MemorySegment.java [hunk 8]
```diff
@@ -897,8 +894,8 @@ static MemorySegment allocateNative(MemoryLayout layout, MemorySession session)
     }
 
     /**
-     * Creates a new native memory segment that models a newly allocated block of off-heap memory with given size (in bytes)
-     * and memory session. A client is responsible make sure that the memory session associated with the returned segment is closed
+     * Creates a native memory segment with given size (in bytes) and memory session.
+     * A client is responsible make sure that the memory session associated with the returned segment is closed
```
Comment (FrauBoes) [inline src/java.base/share/classes/java/lang/foreign/MemorySegment.java:898]: same, "is responsible **for** ensuring that"

File: src/java.base/share/classes/java/lang/foreign/MemorySegment.java [hunk 9]
```diff
@@ -920,10 +917,9 @@ static MemorySegment allocateNative(long bytesSize, MemorySession session) {
     }
 
     /**
-     * Creates a new native memory segment that models a newly allocated block of off-heap memory with given size
-     * (in bytes), alignment constraint (in bytes) and memory session. A client is responsible make sure that the memory
-     * session associated with the returned segment is closed when the segment is no longer in use.
-     * Failure to do so will result in off-heap memory leaks.
+     * Creates a native memory segment with given size (in bytes), alignment constraint (in bytes) and memory session.
+     * A client is responsible make sure that the memory session associated with the returned segment is closed when the
```
Comment (FrauBoes) [inline src/java.base/share/classes/java/lang/foreign/MemorySegment.java:921]: same, "is responsible for ensuring that"

File: src/java.base/share/classes/java/lang/foreign/SequenceLayout.java [hunk 10]
```diff
@@ -98,10 +95,10 @@ public long elementCount() {
     }
 
     /**
-     * Obtains a new sequence layout with same element layout, alignment constraints and name as this sequence layout
-     * but with the new specified element count.
+     * Returns a sequence layout with same element layout, alignment constraints and name as this sequence layout,
```
Comment (FrauBoes) [inline src/java.base/share/classes/java/lang/foreign/SequenceLayout.java:98]: with **the** same element layout

File: src/java.base/share/classes/java/lang/foreign/ValueLayout.java [hunk 11]
```diff
@@ -57,11 +57,8 @@
  * occur. For example, in a future release, synchronization may fail.
  * The {@code equals} method should be used for comparisons.
  *
- * <p> Unless otherwise specified, passing a {@code null} argument, or an array argument containing one or more {@code null}
- * elements to a method in this class causes a {@link NullPointerException NullPointerException} to be thrown. </p>
- *
  * @implSpec
- * This class is immutable and thread-safe.
+ * This class and its subclasses are immutable and thread-safe and <a href="{@docRoot}/java.base/java/lang/doc-files/ValueBased.html">value-based</a>.
```
Comment (FrauBoes) [inline src/java.base/share/classes/java/lang/foreign/ValueLayout.java:61]: immutable, thread-safe and

File: src/java.base/share/classes/java/lang/foreign/ValueLayout.java [hunk 12]
```diff
@@ -92,10 +89,11 @@ public ByteOrder order() {
     }
 
     /**
-     * Returns a new value layout with given byte order.
+     * Returns a value layout with same carrier, alignment constraints and name as this value layout,
```
Comment (FrauBoes) [inline src/java.base/share/classes/java/lang/foreign/ValueLayout.java:92]: with **the** same carrier, ...

**Review by FrauBoes — APPROVED** [review]
Only some minor comments on wording and punctuation.

**Pull request closed without merging**